README.md
# VTuber-Translator-Bot
https://twitter.com/VTuberTransl8r

Twitter bot that translates JP VTuber tweets to English using DeepL translator.

More information coming soon!


v0.2.1


changelog.txt
3/12/22 | v1.0.0
- VTuber Translator Bot full release
- actually fixed issue where specific Tweet lengths would cause the bot to crash (lol)
- added 15 more accounts to the translate list (total of 25)
- new banner and pfp on the Twitter account (by @meebxx)

3/10/22 | v0.2.1
- fixed issue where specific Tweet lengths would cause the bot to crash

3/9/22 | v0.2.0
- embedded media is no longer included in translated replies
- added check to make sure translated Tweet is under 280 character limit
- added check to ensure DeepL character limit is not surpassed
- added 5 more accounts to the translate list (total of 10)
- various small improvements

3/7/22 | v0.1.0
- activated script
- created GitHub repo
- released v0.1.0


rrr.py
import tweepy
import deepl
import numpy as np
from time import sleep

#### Twitter API Keys ####
CONSUMER_KEY = 'INSERT CONSUMER KEY HERE'
CONSUMER_SECRET = 'INSERT CONSUMER SECRET KEY HERE'
ACCESS_KEY = 'INSERT ACCESS KEY HERE'
ACCESS_SECRET = 'INSERT ACCESS SECRET KEY HERE'

auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
api = tweepy.API(auth)
mentions = api.mentions_timeline()

#### DeepL API Key ####
translator = deepl.Translator('INSERT DEEPL KEY HERE')

NAMES_FILE = 'account_names.txt'
ID_FILE = 'last_seen_id.txt'


### Run the main translation ###
def run_translation():
    user_index = 0

    for user in handles_ids[0]:
        handle = user[0]  # index '0' contains handle
        last_tweet = user[1]  # index '1' contains last tweet id

        current_tweet_index = 0

        print('Name: ' + handle + ' ID: ' + last_tweet)
        # tweet_text, tweet_id = get_tweet(handle, current_tweet_index)

        # Find the first tweet that isn't a RT
        while True:
            tweet_text, tweet_id = get_tweet(handle, current_tweet_index)

            if not tweet_text.startswith('RT'):
                break

            print(handle + ' - RT Found')
            current_tweet_index += 1

        # See if last tweet was already translated
        if int(last_tweet) == tweet_id:
            print('Already translated Tweet with ID: ' + last_tweet)
            user_index += 1
            continue

        print(tweet_text)

        translated_text = translate_tweet(tweet_text)

        api.update_status('@' + handle + ' ' + '[AI Translation 🤖]\n\n' + translated_text, in_reply_to_status_id=tweet_id)

        # Update the file to store the new id
        update_last_seen(tweet_id, user_index)
        user_index += 1


### Get most recent tweet number 'index' +1 from set user ###
### Return tweet's text and id ###
def get_tweet(handle, index):
    key = api.user_timeline(screen_name=handle, tweet_mode='extended')  # 'extended' grabs the entire text
    tweet_text = key[index].full_text
    tweet_id = key[index].id
    return tweet_text, tweet_id


### Translate and return? the text ###
def translate_tweet(tweet_text):
    translated = translator.translate_text(tweet_text, source_lang='JA', target_lang='EN-US')
    print(translated)
    tr_string = str(translated)
    print(tr_string)

    return tr_string


### Initialize the main list from files (List = [[HANDLE, ID], [HANDLE, ID]...]) ###
def initialize_list():
    name_file = open('account_names.txt', 'r')
    id_file = open('last_seen_id.txt', 'r')

    names = []
    ids = []

    # Add names and ids from respective files until there are none left
    while True:
        name_line = name_file.readline()
        id_line = id_file.readline()

        if not name_line:
            break

        names.append(name_line.strip())
        ids.append(id_line.strip())

    name_file.close()
    id_file.close()

    return np.dstack((names, ids))


### Update last_seen_id.txt to the given id ###
def update_last_seen(tweet_id, index):
    # Read file into an array
    with open('last_seen_id.txt', 'r') as file:
        data = file.readlines()

    # Update the user's most recent id in the array
    data[index] = str(tweet_id) + '\n'

    # Write everything back
    with open('last_seen_id.txt', 'w') as file:
        file.writelines(data)

    file.close()


while True:
    handles_ids = initialize_list()
    print(handles_ids)
    run_translation()
    # Wait 15 minutes before running again
    sleep(900)


translator_bot.py
import tweepy
import deepl
import numpy as np
from time import sleep

#### Twitter API Keys ####
CONSUMER_KEY = 'INSERT CONSUMER KEY HERE'
CONSUMER_SECRET = 'INSERT CONSUMER SECRET KEY HERE'
ACCESS_KEY = 'INSERT ACCESS KEY HERE'
ACCESS_SECRET = 'INSERT ACCESS SECRET KEY HERE'

auth = tweepy.OAuthHandler(CONSUMER_KEY, CONSUMER_SECRET)
auth.set_access_token(ACCESS_KEY, ACCESS_SECRET)
api = tweepy.API(auth)
mentions = api.mentions_timeline()

#### DeepL API Key ####
translator = deepl.Translator('INSERT DEEPL KEY HERE')

NAMES_FILE = 'account_names.txt'
ID_FILE = 'last_seen_id.txt'


### Run the main translation ###
def run_translation():
    user_index = 0

    for user in handles_ids[0]:
        handle = user[0]  # index '0' contains handle
        last_tweet = user[1]  # index '1' contains last tweet id

        current_tweet_index = 0

        print('Name: ' + handle + ' ID: ' + last_tweet)

        # Find the first tweet that isn't a RT
        while True:
            tweet_text, tweet_id = get_tweet(handle, current_tweet_index)

            if not tweet_text.startswith('RT'):
                break

            print(handle + ' - RT Found')
            current_tweet_index += 1

        # See if last tweet was already translated
        if int(last_tweet) == tweet_id:
            print('Already translated Tweet with ID: ' + last_tweet)
            user_index += 1
            continue

        # Remove links to any media in the Tweet
        if "https://" in tweet_text:  # NOTE: This could potentially break if there is a standard link. I'll change this if it becomes an issue
            tweet_text = tweet_text[:tweet_text.index("https://")]

        print(tweet_text)

        # Check if enough characters remain to translate the Tweet
        if check_remaining_characters(len(tweet_text)):
            continue

        translated_text = translate_tweet(tweet_text)

        # Make sure translated Tweet is under 280 character limit
        tweet_length = count_characters(translated_text)
        print(tweet_length)
        if tweet_length > 280:
            print('Tweet was too long (must be less than or equal to 280 characters; Tweet was ' + str(tweet_length) + ')')
            update_last_seen(tweet_id, user_index)
            user_index += 1
            continue

        api.update_status('@' + handle + ' ' + '【AI Translation 🤖】\n\n' + translated_text, in_reply_to_status_id=tweet_id)

        # Update the file to store the new id
        update_last_seen(tweet_id, user_index)
        user_index += 1


### Get most recent tweet number 'index' +1 from set user ###
### Return tweet's text and id ###
def get_tweet(handle, index):
    key = api.user_timeline(screen_name=handle, tweet_mode='extended')  # 'extended' grabs the entire text
    tweet_text = key[index].full_text
    tweet_id = key[index].id
    return tweet_text, tweet_id


### Translate and return? the text ###
def translate_tweet(tweet_text):
    translated = translator.translate_text(tweet_text, source_lang='JA', target_lang='EN-US')
    tr_string = str(translated)
    print(tr_string)

    return tr_string


### Initialize the main list from files (List = [[HANDLE, ID], [HANDLE, ID]...]) ###
def initialize_list():
    name_file = open('account_names.txt', 'r')
    id_file = open('last_seen_id.txt', 'r')

    names = []
    ids = []

    # Add names and ids from respective files until there are none left
    while True:
        name_line = name_file.readline()
        id_line = id_file.readline()

        if not name_line:
            break

        names.append(name_line.strip())
        ids.append(id_line.strip())

    name_file.close()
    id_file.close()

    return np.dstack((names, ids))


### Update last_seen_id.txt to the given id ###
def update_last_seen(tweet_id, index):
    # Read file into an array
    with open('last_seen_id.txt', 'r') as file:
        data = file.readlines()

    # Update the user's most recent id in the array
    data[index] = str(tweet_id) + '\n'

    # Write everything back
    with open('last_seen_id.txt', 'w') as file:
        file.writelines(data)

    file.close()


### Make sure there are enough characters remaining for translation ###
def check_remaining_characters(tweet_len):
    usage = translator.get_usage()

    if tweet_len + usage.character.count > usage.character.limit:
        print("Unable to translate. Out of characters.")
        return True

    return False


### Count the number of characters in a Tweet ###
def count_characters(tweet):
    numChars = 0
    for char in tweet:
        charASCII = ord(char)
        if 32 <= charASCII <= 126 or charASCII == 10:   # ASCII 10 is newline
            numChars += 1
        else:
            numChars += 2  # Emojis and other special characters count as 2 on Twitter

    return numChars + len('【AI Translation 🤖】') + 5


while True:
    handles_ids = initialize_list()
    print(handles_ids)
    run_translation()
    # Wait 15 minutes before running again
    sleep(900)
